### PR TITLE
[compiler] use reactiveConditionHelper to reduce size

### PR DESCRIPTION
**Pull request by GrinZero**
I found that when studying react compiler, the compiler always compiles `scope` into very similar conditional comparison code blocks. So I am thinking about whether it is possible to extract these similar code blocks into a common function, which may help reduce the code volume of large-scale projects.

## Overview

This PR introduces a new runtime, which I call conditionHelper. It organizes reusable logic in the scope and separates it into common functions to reduce the volume of compiled code. Although this is only an initial implementation and there are still testing and other tasks to be completed, it has a certain positive effect on reducing the volume of compiled code.

## Demo

Modifying the babel option allows you to experience helpers, like:
```js
...
  babel: {
    plugins: [
      [
        "babel-plugin-react-compiler",
        {
          environment: {
            enableReactiveScopeConditionHelper: true,
          },
        },
      ],
    ],
  },
...
```

And run `build` or `dev`:
image

TIP: This is my first time submitting a contribution to React. If anything that needs to be done is missing, please remind me and I will handle it as soon as possible

### CODE REVIEW

**Comment by GrinZero** [comment]
cc @sebmarkbage or anybody? I hope this PR can catch everyone's attention

**Review by josephsavona — COMMENTED** [review]
Thanks for the PR. We're not opposed to exploring changes to the output format, but:
a) We want to find one format that is the best combination of code size and runtime performance, not have long-lived feature flags for people to choose.
b) If we use higher-level helpers, the cost of those has to be carefully tested to ensure they don't regress other properties. 

In this case, I'm concerned that this reintroduces lots of array and closure allocations in exchange for reduced code size. So far we've been mostly memory neutral, but this could easily make the compiled output use more memory than before. 

If you finish implementing this such that CI is green, we can try benchmarking it. But we'd only be move forward if the results are positive.

File: compiler/packages/react-compiler-runtime/src/index.ts [hunk 1]
```diff
@@ -44,6 +44,41 @@ export function $read(memoCache: MemoCache, index: number) {
...
+      ? (i: number) => cache[conditionIndexs[i]] === $empty
+      : (i: number) => cache[conditionIndexs[i]] !== conditionTargets[i];
+  const condition = (() => {
+    for (let i = 0; i < conditionIndexs.length; i++) {
+      if (test(i)) {
+        return true;
+      }
+    }
+    return false;
+  })();
```
Comment (josephsavona) [inline compiler/packages/react-compiler-runtime/src/index.ts:66]: this is really inefficient. Let's turn this into static code without the unnecessary closure allocations. ie

```js
let hasChange = false;
for (let i = 0; i < conditionIndices.length; i++) {
   ...
   if (cache[conditionIndices[i]] !== conditionTargets[i]) {
     hasChange = true;
     break;
   }
   ...
}
```

**Review by josephsavona — CHANGES_REQUESTED** [review]
requesting changes to get this to pass CI before we can proceed with anything

**Pull request closed without merging**